### PR TITLE
drivers: Extend coverage for UARTE driver

### PR DESCRIPTION
**Pull request by nordic-bami**
Add more test cases for UARTE driver API
Signed-off-by: Bartosz Miller <bartosz.miller@nordicsemi.no>

### CODE REVIEW

File: tests/drivers/uart/uart_elementary/README.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+The purpose of this test is to validate basic UART driver functions,
+that are not tested elsewhere.
+UART interrupt mode is used for the tests purpose.
+
+Hardware setup required for these tests:
+The following pins need to be connected together:
+ - P0.02 -- P0.04
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/README.txt:7]: this is board specific and should not be here.

File: tests/drivers/uart/uart_elementary/boards/nrf54l15pdk_nrf54l15_cpuapp_dual_uart.overlay [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+dut_aux: &uart22 {
+	status = "okay";
+	current-speed = <115200>;
+	pinctrl-0 = <&uart22_default>;
+	pinctrl-1 = <&uart22_sleep>;
+	pinctrl-names = "default", "sleep";
+};
+
+&grtc {
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf54l15pdk_nrf54l15_cpuapp_dual_uart.overlay:51]: can be removed. Looks like copy-paste from uart_mix_fifo_poll test where it was used.

File: tests/drivers/uart/uart_elementary/boards/nrf54l15pdk_nrf54l15_cpuapp.overlay [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+dut: &uart21 {
+	status = "okay";
+	current-speed = <115200>;
+	pinctrl-0 = <&uart21_default>;
+	pinctrl-1 = <&uart21_sleep>;
+	pinctrl-names = "default", "sleep";
+	hw-flow-control;
+};
+
+&grtc {
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf54l15pdk_nrf54l15_cpuapp.overlay:33]: can be removed. Looks like copy-paste from uart_mix_fifo_poll test where it was used.

File: tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	pinctrl-names = "default", "sleep";
+	current-speed = <115200>;
+};
+
+&dut_aux {
+	memory-regions = <&cpuapp_dma_region>;
+};
+
+
+&grtc {
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay:65]: can be removed. Looks like copy-paste from uart_mix_fifo_poll test where it was used.

File: tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp.overlay [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "nrf54h20dk_nrf54h20_common.dtsi"
+
+
+&dut {
+	memory-regions = <&cpuapp_dma_region>;
+};
+
+&grtc {
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:10]: can be removed. Looks like copy-paste from uart_mix_fifo_poll test where it was used.

File: tests/drivers/uart/uart_elementary/boards/nrf5340dk_nrf5340_cpuapp.overlay [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	compatible = "nordic,nrf-uarte";
+	status = "okay";
+	pinctrl-0 = <&uart1_default_alt>;
+	pinctrl-1 = <&uart1_sleep_alt>;
+	pinctrl-names = "default", "sleep";
+	hw-flow-control;
+};
+
+&rtc1 {
+	interrupts = <21 2>;
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf5340dk_nrf5340_cpuapp.overlay:35]: can be removed. Looks like copy-paste from uart_mix_fifo_poll test where it was used.

File: tests/drivers/uart/uart_elementary/boards/nrf5340dk_nrf5340_cpuapp.overlay [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+};
+
+counter_dev: &timer0 {
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf5340dk_nrf5340_cpuapp.overlay:38]: can be removed. Looks like copy-paste from uart_mix_fifo_poll test where it was used.

File: tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+dut_aux: &uart137 {
+	status = "okay";
+	pinctrl-0 = <&uart137_default_alt>;
+	pinctrl-1 = <&uart137_sleep_alt>;
+	pinctrl-names = "default", "sleep";
+	current-speed = <115200>;
+};
+
+&dut_aux {
+	memory-regions = <&cpuapp_dma_region>;
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay:61]: can it be moved to the node above?

File: tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay [hunk 9]
```diff
@@ -0,0 +1,67 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "nrf54h20dk_nrf54h20_common.dtsi"
+
+
+&dut {
+	memory-regions = <&cpuapp_dma_region>;
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay:7]: can it be moved to the dut uart135 node?

File: tests/drivers/uart/uart_elementary/testcase.yaml [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+tests:
+  drivers.uart.uart_elementary:
+    filter: CONFIG_SERIAL_SUPPORT_INTERRUPT
+    platform_allow:
+      - nrf54h20dk/nrf54h20/cpuapp
+      - nrf54l15pdk/nrf54l15/cpuapp
+      - nrf5340dk/nrf5340/cpuapp
+    extra_configs:
+      - CONFIG_UART_INTERRUPT_DRIVEN=y
+      - CONFIG_UART_0_INTERRUPT_DRIVEN=y
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/testcase.yaml:16]: nrf54s don't have instance 0. Shouldn't those 2 entries be removed?

File: tests/drivers/uart/uart_elementary/testcase.yaml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+      - CONFIG_UART_0_ENHANCED_POLL_OUT=n 
+      - CONFIG_UART_NRFX_UARTE_LEGACY_SHIM=n
+  drivers.uart.uart_elementary_dual_nrf54h:
+    filter: CONFIG_SERIAL_SUPPORT_INTERRUPT
+    platform_allow:
+      - nrf54h20dk/nrf54h20/cpuapp
+    extra_args: DTC_OVERLAY_FILE="boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay"
+    extra_configs:
+      - CONFIG_UART_INTERRUPT_DRIVEN=y
+      - CONFIG_UART_0_INTERRUPT_DRIVEN=y
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/testcase.yaml:26]: configs for instance 0 and 1 can be removed. Is asynchronous  API is not enabled then enabling globally interrupt driven API will enable it for all instances.

File: tests/drivers/uart/uart_elementary/testcase.yaml [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      - CONFIG_DUAL_UART_TEST=y
+      - CONFIG_UART_NRFX_UARTE_LEGACY_SHIM=n
+  drivers.uart.uart_elementary_dual_setup_mismatch_nrf54h:
+    filter: CONFIG_SERIAL_SUPPORT_INTERRUPT
+    platform_allow:       
+      - nrf54h20dk/nrf54h20/cpuapp
+    extra_args: DTC_OVERLAY_FILE="boards/nrf54h20dk_nrf54h20_cpuapp_dual_uart.overlay"
+    extra_configs:
+      - CONFIG_UART_INTERRUPT_DRIVEN=y
+      - CONFIG_UART_0_INTERRUPT_DRIVEN=y
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/testcase.yaml:39]: ditto

File: tests/drivers/uart/uart_elementary/testcase.yaml [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      - CONFIG_SETUP_MISMATCH_TEST=y
+      - CONFIG_UART_NRFX_UARTE_LEGACY_SHIM=n
+  drivers.uart.uart_elementary_dual_nrf54l:
+    filter: CONFIG_SERIAL_SUPPORT_INTERRUPT
+    platform_allow: 
+      - nrf54l15pdk/nrf54l15/cpuapp
+    extra_args: DTC_OVERLAY_FILE="boards/nrf54l15pdk_nrf54l15_cpuapp_dual_uart.overlay"
+    extra_configs:
+      - CONFIG_UART_INTERRUPT_DRIVEN=y
+      - CONFIG_UART_0_INTERRUPT_DRIVEN=y
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/testcase.yaml:53]: ditto

**Review by nordic-krch — COMMENTED** [review]
Few comments. Just looked into configuration files so far.

File: tests/drivers/uart/uart_elementary/testcase.yaml (hunk 10 above)
Comment (nordic-bami) [reply tests/drivers/uart/uart_elementary/testcase.yaml:16]: How to enable interrupt driven UART api for nrf54h with K-configs ?

File: tests/drivers/uart/uart_elementary/src/main.c [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+
+#if defined(CONFIG_SETUP_MISMATCH_TEST)
+volatile uint8_t uart_error_counter = 0;
+#endif
+
+#if defined(CONFIG_DUAL_UART_TEST)
+static const struct device *const uart_dev_aux = DEVICE_DT_GET(UART_NODE_AUX);
+uint8_t test_buffer_aux[TEST_BUFFER_LEN] = { 0x00, 0x00, 0x00, 0x00, 0x00,
+					     0x00, 0x00, 0x00, 0x00, 0x00 };
+volatile uint8_t aux_uart_error = 0;
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/src/main.c:51]: use `static` for all variables here.

File: tests/drivers/uart/uart_elementary/prj.conf [hunk 15]
```diff
@@ -0,0 +1,7 @@
+CONFIG_SERIAL=y
+CONFIG_ZTEST=y
+CONFIG_TEST_USERSPACE=y
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_UART_USE_RUNTIME_CONFIGURE=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_UART_NRFX_UARTE_LEGACY_SHIM=n
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/prj.conf:7]: it's nordic specific Kconfig and test is generic so it shouldn't be here. It's the default option for nrf54l/h so it's only have to be forced for nrf53 (maybe board specific conf file?).

File: tests/drivers/uart/uart_elementary/src/main.c [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+
+#define SLEEP_TIME_US 1000
+#define TEST_BUFFER_LEN 10
+
+static const struct device *const uart_dev = DEVICE_DT_GET(UART_NODE);
+
+const uint8_t test_pattern[TEST_BUFFER_LEN] = { 0x11, 0x12, 0x13, 0x14, 0x15,
+						0x16, 0x17, 0x18, 0x19, 0x20 };
+uint8_t test_buffer[TEST_BUFFER_LEN] = {
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/src/main.c:38]: static variables are by default zeroed so it's redundant. same for `test_buffer_aux`.

File: tests/drivers/uart/uart_elementary/src/main.c [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+Callback function for AUX UART interrupt based transmission test
+*/
+static void interrupt_driven_uart_callback_aux_uart(const struct device *dev, void *user_data)
+{
+	int err;
+	static int tx_byte_offset_aux = 0;
+	static int rx_byte_offset_aux = 0;
+	uart_irq_update(dev);
+	err = uart_err_check(dev);
+#if !defined(CONFIG_COVERAGE)
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/src/main.c:127]: can you add comment why it is needed?

File: tests/drivers/uart/uart_elementary/src/main.c [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+	while (uart_irq_is_pending(dev)) {
+		if (uart_irq_rx_ready(dev)) {
+			uart_rx_interrupt_service(dev, (uint8_t *)user_data, &rx_byte_offset_aux);
+		}
+		if (uart_irq_tx_ready(dev)) {
+			uart_tx_interrupt_service(dev, &tx_byte_offset_aux);
+		}
+	}
+}
+#endif
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/src/main.c:139]: please add in a comment a reference to starting #if statement `/* CONFIG_DUAL_UART_TEST */`

File: tests/drivers/uart/uart_elementary/src/main.c [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+		zassert_equal(test_buffer[index], test_pattern[index],
+			      "UART0 recieived data byte %d does not match pattern 0x%x != 0x%x",
+			      index, test_buffer[index], test_pattern[index]);
+		zassert_equal(test_buffer_aux[index], test_pattern[index],
+			      "UART1 recieived data byte %d does not match pattern 0x%x != 0x%x",
+			      index, test_buffer_aux[index], test_pattern[index]);
+	}
+#endif
+}
+#endif
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_elementary/src/main.c:314]: add a comment about #if statement.

**Review by nordic-segl — APPROVED** [review]

**Review by nordic-piks — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**